File: leetcode/easy/1700-sandwiches/queue.py
"""
  @author Vortexx2
  Problem 1700 - Number of students unable to eat lunch

  Runtime - 16 ms
  Memory Usage - 14.3 MB
"""
from typing import List
from collections import deque


class Solution:
  def countStudents(self, students: List[int], sandwiches: List[int]) -> int:
    """
      @approach Go through students once and strike off all sandwiches that can be eaten in the first go. If student can't eat sandwich, push index to queue. Then -> for the length of queue, try to strike of sandwiches. If after one iteration over length of queue length does not decrease, we have reached the end.
    """
    dq = deque([])
    i, j = 0, 0

    while i < len(students):
      if j == len(sandwiches):
        return 0

      if students[i] == sandwiches[j]:
        j += 1

      else:
        dq.appendleft(i)

      i += 1

    count = len(dq)
    while count:

      for i in range(count):
        if students[dq[-1]] == sandwiches[j]:
          dq.pop()
          j += 1

        else:
          dq.appendleft(dq.pop())

      if count == len(dq):
        return count

      count = len(dq)

    return 0


if __name__ == "__main__":
  sol = Solution()
  print(sol.countStudents([1, 1, 0, 0], [0, 1, 0, 1]))

File: leetcode/easy/1700-sandwiches/queue.cpp
/**
 * @file queue.cpp
 * @author Vortexx2
 * @brief Problem 1700 - Number of Students Unable to Eat Lunch
 * @date 20-07-2021
 * 
 * Runtime - 0 ms
 * Memory Usage - 8.6 MB
 */
#include <iostream>
#include <vector>
#include <deque>
#include <unordered_map>
#include <map>
#include <set>
#include <stack>
#include <queue>
#include <string>

using namespace std;

class Solution
{
public:
  int countStudents(vector<int> &students, vector<int> &sandwiches)
  {
    int i = 0, j = 0;
    queue<int> q;

    while (i < students.size())
    {
      if (j == sandwiches.size())
        return 0;

      if (students[i] == sandwiches[j])
        j++;

      else
        q.push(i);

      i++;
    }

    int count = q.size();
    while (count)
    {

      for (int k = 0; k < count; k++)
      {
        if (students[q.front()] == sandwiches[j])
        {
          j++;
          q.pop();
        }
        else
        {
          q.push(q.front());
          q.pop();
        }
      }

      if (count == q.size())
        return q.size();

      count = q.size();
    }

    return 0;
  }
};